pkg/querier/chunk_merge_iterator.go

```diff
@@ -153,3 +157,177 @@ type byFrom []*chunkIterator
 func (b byFrom) Len() int           { return len(b) }
 func (b byFrom) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
 func (b byFrom) Less(i, j int) bool { return b[i].From < b[j].From }
+
+type nonOverlappingIterator struct {
+	its []*chunkIterator
+
+	curr *chunkIterator
+}
+
+// newNonOverlappingIterator returns a single iterator over an slice of sorted,
+// non-overlapping iterators.
+func newNonOverlappingIterator(its []*chunkIterator) *nonOverlappingIterator {
+	return &nonOverlappingIterator{
+		its: its,
+	}
+}
+
+func (it *nonOverlappingIterator) Seek(t int64) bool {
+	for it.curr == nil || !it.curr.Seek(t) {
+		if len(it.its) == 0 {
+			return false
+		}
+
+		it.curr = it.its[0]
+		it.its = it.its[1:]
+	}
+
+	return true
+}
+
+func (it *nonOverlappingIterator) Next() bool {
+	for it.curr == nil || !it.curr.Next() {
+		if len(it.its) == 0 {
+			return false
+		}
+
+		it.curr = it.its[0]
+		it.its = it.its[1:]
+	}
+
+	return true
+}
+
+func (it *nonOverlappingIterator) At() (int64, float64) {
+	return it.curr.At()
+}
+
+func (it *nonOverlappingIterator) Err() error {
+	if it.curr == nil {
+		return nil
+	}
+
+	return it.curr.Err()
+}
+
+type chunkMergeIteratorV2 struct {
+	its []*nonOverlappingIterator
+	h   seriesIteratorHeap
+
+	curr    *nonOverlappingIterator
+	lastErr error
+}
+
+func newChunkMergeIteratorV2(cs []chunk.Chunk) storage.SeriesIterator {
+	chunks := make([]*chunkIterator, len(cs))
+	for i := range cs {
+		chunks[i] = &chunkIterator{
+			Chunk: cs[i],
+			it:    cs[i].Data.NewIterator(),
+		}
+	}
+	sort.Sort(byFrom(chunks))
+
+	chkLists := [][]*chunkIterator{}
+Outer:
+	for _, chk := range chunks {
+		for i, chkList := range chkLists {
+			if chkList[len(chkList)-1].Through.Before(chk.From) {
+				chkLists[i] = append(chkLists[i], chk)
+				continue Outer
+			}
+		}
+
+		// Here only if there is no list that accepted this chunk. Need a new one.
+		chkLists = append(chkLists, []*chunkIterator{chk})
+	}
+
+	its := make([]*nonOverlappingIterator, 0, len(chkLists))
+	for _, chkList := range chkLists {
+		its = append(its, newNonOverlappingIterator(chkList))
+	}
+
+	c := &chunkMergeIteratorV2{
```

Comment:
We should reduce duplication and adapt chunkMergeIterator so we can return it here.